src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -0,0 +1,350 @@
+package com.fulcrumgenomics.umi
+
+import java.nio.file.Paths
+import java.util.concurrent.atomic.AtomicLong
+
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.{Io, ProgressLogger, Sequences, SimpleCounter}
+import com.fulcrumgenomics.util.Sequences.countMismatches
+import dagr.commons.CommonsDef.{DirPath, PathToBam}
+import dagr.commons.util.LazyLogging
+import dagr.sopt.cmdline.ValidationException
+import dagr.sopt.{arg, clp}
+import htsjdk.samtools.SAMFileHeader.{GroupOrder, SortOrder}
+import htsjdk.samtools._
+import htsjdk.samtools.util.SortingCollection
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+
+object GroupReadsByUmi {
+  private val SortPositionAttribute        = "__GRs"
+  private val LibraryAttribute             = "__GRlb"
+  private val InsertEndsAndStrandAttribute = "__Insert"
+
+  /** Trait that can be implemented to provide a UMI assignment strategy. */
+  sealed trait UmiAssigner {
+    def assign(rawUmis: Seq[String]) : Map[String, String]
+  }
+
+  /** Assigns UMIs based solely on sequence identity. */
+  class IdentityUmiAssigner extends UmiAssigner {
+    override def assign(rawUmis: Seq[String]): Map[String, String] = rawUmis.map(umi => umi -> umi.toUpperCase).toMap
+  }
+
+  /** Assigns UMIs based solely on sequence identity. */
+  class SimpleErrorUmiAssigner(val maxMismatches: Int) extends UmiAssigner {
+    override def assign(rawUmis: Seq[String]): Map[String, String] = {
+      type UmiSet = mutable.SortedSet[String]
+      val umiSets = ListBuffer[UmiSet]()
+      for (umi <- rawUmis) {
+        val matchedSets = mutable.Set[UmiSet]()
+        for (set <- umiSets) {
+          if (set.exists(other => countMismatches(other, umi) <= maxMismatches)) {
+            set.add(umi)
+            matchedSets.add(set)
+          }
```

Comment:
How wedded are we to wanting to use `map/filter/foreach` versus a nice loop like you have, that to me reads better?
```umiSets.filter {set => set.exists(other => countMismatches(other, umi) <= maxMismatches }.foreach { set => { set.add(umi); matchSets.add(set) }}```?